# Methylation-
This document aims to simplify the future processing of bisulfite data in various methylation studies. The main steps in this pipeline are    . 

# genome annotation - Synteny - Ds computation - changepoint analysis - whole genome alignments 
====================================================================================

   * [Purpose](#purpose)
   * [Dependencies](#dependencies)
   * [Input data](#input-data)
   * [Example input data](#example-input-data)
   * [Quick start](#quick-start)
   * [Example plot](#example-plot)
   * [Detailed steps](#detailed-steps)


# Purpose:
##  sets of scripts to : 
 1 - Trim (primer extraction), 

 2 - Maps (mapping of reads to the species reference sequence) 

 3 - Annotation and compartmentalisation  

 4 - Quantify (quantification of methylation levels at different genome scales) 





# Dependencies: 

basic requirements: **DNMTools**, **gcc** , **R**, **make**, **cmake**, **wget** or **curl** , **java** 

conda or **mamba** can be used to install several dependencies, especially without root access 

### First thing is to get mamba or conda. 

I recommand mamba for linux: 

```
curl -L -O https://github.com/conda-forge/miniforge/releases/latest/download/Miniforge-pypy3-Linux-x86_64.sh
bash Miniforge-pypy3-Linux-x86_64.sh
#note: see here for other verions of mamba: https://github.com/conda-forge/miniforge
```


# HOW TO USE:   

 
# Input data: 


* minimal input by species:
	* Bisulfite data [generated by the Novogène laboratory from extrated DNA] (**fastq files**) 
    * genomes assemblies (**single fasta file**)
 
 <img src="https://raw.githubusercontent.com/PaulineMichelZoe/Methylation-/main/Biseq-novogene.jpg" width="490" height="490">


We used Dnmtools software to map these reads onto the reference genome. Here We are obliged to use a third file as an adapter because the reference genomes and the sequences processed by novogenes had different names for the same species. The first step is to remove the primers that have also been amplified with abismal:
```
On a pas le Trim (voir Ricardo)
    awk '{for(i=2;i<=NF;i++) if($i!~/NA/) print $1,$i}' Documents/data/ref2bseq.txt | sed 's/B[0-9]://g;s/_bis/-bis/g;s/.fa /.idx /1' | awk '{split($NF,r,"_"); print $1,r[1]}' | awk '!s[$2]++' | sort -k2,2 | join -1 2 -2 1 - <(locate val_1.fq.gz | sed 's/_bis/-bis/g' | awk -F "/" '!/Trash/ {split($NF,s,"_"); print s[1],$0}' | awk '{gsub(/-bis/,"_bis",$NF); print}' | sort -k1,1) | awk '{print "abismal -t 10 -i /home/pauline/Documents/data/"$2" -s /home/pauline/Documents/data/mapstat/"$1"-"NR".mapstat -o /home/pauline/Documents/data/new/"$1"-"NR".sam "$NF,$NF}' | sed 's/1_val_1/2_val_2/2' | grep -f <(for i in $(locate val_1.fq.gz | awk -F "/" '!/Trash/ && !s[$NF]++'); do ls -lrt $i; done | awk '/mars/ && $(NF-2)==16 {print $NF}' | sort -u ) - |
```
The second step is to map the different reads to the reference genome. If the result of the mapping shows a similarity of less than 15% to the reference genome, it is considered to be a problem of contamination or poor DNA extraction and the replicas will not be kept for future analysis. 
```  
  awk '{gsub(/_bis/,"-bis",$0);  for(i=2;i<=NF;i++) if($i!~/NA/) print $1,$i}' Documents/data/ref2bseq.txt | awk '{split($NF,r,"_"); print $1,r[1]}' | awk '!s[$2]++ {sub(/^B[12]:/,"",$2); print}' | sort -k2,2 | join -1 2 -2 1 - <(find /home/pauline/Documents/data/new_data/ -name "*sorted.Sam-out-sorted.Sam" | sed 's/-bis/_bis/1' | awk -F "/" '{split($NF,p,"-"); print p[1],$0}' | sed 's/_bis/-bis/1'  | sort -k1,1) | awk '{print "dnmtools counts -c Documents/data/seqref/"$2"  -o Documents/data/new_data/"$1"-"NR".meth  "$3}'  | bash 
``` 
 ``` 
 find /home/pauline/Documents/data/hypermeth/ -name "*merge.meth"  | awk -F "/" '!/sorted/{split($NF,f,"."); print "dnmtools sym -m -o  /home/pauline/Documents/data/hypermeth/"f[1]"-CpG-merge.meth "$0}'| bash)
 ```



# Quick start:

***before running the pipeline make sure to have ALL dependencies installed***

***make sure to have the correct input data***

### step1 - edit the config file and set the correct path 

the config file is here: `./config/config`





# detailed steps  


# list of operations and tools


| __Operation__                     |  __Tools__                         |  __data type__  | 
|:---------------------------------:|:------------------------------:|:-----------:| 
| __read trimming__                |  Trimgalore                   | SeqBisulfite         | 
| __read mapping__                 |  abismal                    | SeqBisulfite          | 
| __sorting read__                 |  samtools                      | SeqBisulfite        |
| __mapping quality assement__     |  samtools + R                  | SeqBisulfite       |




This code has been tested with linux. 





